File: modules/core/app/model/exercisereading/ExerciseCollectionReader.java
package model.exercisereading;

import io.ebean.Finder;
import model.exercise.Exercise;
import model.exercise.ExerciseCollection;

public abstract class ExerciseCollectionReader<E extends Exercise, C extends ExerciseCollection<E>>
    extends JsonReader<C> {

  protected ExerciseReader<E> delegateReader;

  public ExerciseCollectionReader(String theExerciseType, ExerciseReader<E> theDelegateReader,
      Finder<Integer, C> theFinder, Class<?> theClassFor) {
    super(theExerciseType, theFinder, theClassFor);
    delegateReader = theDelegateReader;
  }

  public ExerciseReader<E> getDelegateReader() {
    return delegateReader;
  }
  
  public void saveExercise(C exercise) {
    exercise.save();
    exercise.getExercises().forEach(Exercise::save);
  }
}

File: modules/core/app/model/exercisereading/ExerciseCollectionReader.scala
package model.exercisereading

import scala.collection.JavaConverters.asScalaBufferConverter

import com.fasterxml.jackson.databind.JsonNode

import io.ebean.Finder
import model.StringConsts.{ AUTHOR_NAME, TEXT_NAME, TITLE_NAME }
import model.exercise.{ Exercise, ExerciseCollection }

abstract class ExerciseCollectionReader[E <: Exercise, C <: ExerciseCollection[E]](et: String, f: Finder[Integer, C], cf: Class[_])
  extends JsonReader[C](et, f, cf) {

  override def save(collection: C) {
    collection.save()
    collection.getExercises().asScala.foreach(_.save())
  }

  def update(coll: C, node: JsonNode) {
    coll.title = node.get(TITLE_NAME).asText
    coll.author = node.get(AUTHOR_NAME).asText
    coll.text = JsonReader.readTextArray(node.get(TEXT_NAME)).mkString("")
  }

}
